pkg/kubelet/cm/helpers.go

```diff
@@ -76,3 +78,20 @@ func buildContainerMapAndRunningSetFromRuntime(ctx context.Context, runtimeServi
 	}
 	return containerMap, runningSet
 }
+
+// GetAllocatableMemory returns memory capacity with out hugePage capacity.
+func GetAllocatableMemory(capacity v1.ResourceList) resource.Quantity {
+	allocatableMemory := capacity[v1.ResourceMemory].DeepCopy()
+	// for every huge page reservation, we need to remove it from allocatable memory
+	for k, v := range capacity {
+		if v1helper.IsHugePageResourceName(k) {
+			value := v.DeepCopy()
+			allocatableMemory.Sub(value)
+			if allocatableMemory.Sign() < 0 {
+				// Negative Allocatable resources don't make sense.
+				allocatableMemory.Set(0)
+			}
+		}
+	}
+	return allocatableMemory
+}
```

Comment:
I see that we are removing memory corresponding to hugepages from the internal memory capacity and I understand what has motivated this but I can't help but wonder why this was added in the first place?

Referring to `CapacityFromMachineInfo` I can see that memory and hugepages are enumerated separately in cadvisor so why we end up with hugepage memory being accounted for under memory is not clear to me: https://github.com/kubernetes/kubernetes/blob/f24316a7b829e619a40cce6cb6b622cb4a3dca50/pkg/kubelet/cadvisor/util.go#L34-L54


IOW, I think we should look into why cadvisor provides information in this manner in the first place, whether this is an intentional behavior or an unintended consequence of something that we don't understand right now. Let's do our due diligence and identify if there is a strong reason for not fixing this at cadvisor level before we take this path.

Reply:
@lance5890 please address this review comment, thanks.

Reply:
I revisited this while trying to write the e2e, and agree with @swatisehgal. Will bring this up in tomorrow's Sig Node CI meeting to finalize the fix here. I want to understand why capacity from machine info takes Huge pages into account. cc: @yujuhong

Reply:
Also refer: https://github.com/kubernetes/kubernetes/pull/99943#issuecomment-826303348